README.md

```diff
@@ -1,71 +1,157 @@
+ko apply -f example-policies
 ```
 
-`Upstream` will allow for you to [sync your fork](https://help.github.com/articles/syncing-a-fork/)
+Note that to actually use a built-in policy in a rollout, the relevant Service or Configuration must have the appropriate annotations to specify the name of the policy that it is using.
 
-### Building
+### Deploying your service
 
-`go build ./...`
+Suppose you have a Knative Service `default/my-app` in a file called `my-app.yaml`, then you can simply run the following command to deploy it:
 
-### Running unit tests
+```sh
+kubectl apply -f my-app.yaml
+```
 
-`go test ./...`
+If this is the first Revision of `my-app`, then the behavior is the same as if KCD is not present. However, if this is not the first Revision of `my-app`, and if `my-app.yaml` correctly identifies a rollout policy in the annotations, then KCD will automatically compute the correct routing state for `my-app` and initiate a progressive rollout. You can watch the traffic percentages in real time via the following command:
 
-### Updating dependencies
+```sh
+kubectl get -w route my-app -o yaml
+```
+
+## Common Failures & Diagnostics
+
+### Incorrect traffic behavior
+
+If you see that the rollout percentage for a Revision you just created stays at 0 or jumps to 100 directly, there can be a few potential culprits:
 
-`./hack/update-deps.sh`
+1. You have not applied any rollout policy.
+2. You have not annotated your Service/Configuration with the correct policy name.
+3. You are not using Knative Service, and your Route and Configuration have namespace and/or name mismatches.
+
+### Debugging tips
+
+You can view the logs of the KCD controller and/or webhook by using the following command:
+
+```sh
+kubectl logs -n knative-serving <pod_name_for_controller/webhook>
+```
 
-## Source Code Headers
+## Under-the-Hood Technicalities
 
-Every file containing source code must include copyright and license
-information. This includes any JS/CSS files that you might be serving out to
-browsers. (This is to help well-intentioned people avoid accidental copying that
-doesn't comply with the license.)
+### Continuous Delivery Webhook
 
-Apache header:
+The continuous delivery webhook serves two purposes:
 
-    Copyright 2020 Google LLC
+1. To make sure that the KService reconciler does not overwrite changes to the Route object, the webhook implements a method that intercepts and overwrites the Route with the desired routing spec before the Route reaches the K8s API server.
+2. To make sure that customized policies strictly adhere to all the required criteria, the webhook implements a method that valids any Policy object it receives, and rejects any errors.
 
-    Licensed under the Apache License, Version 2.0 (the "License");
-    you may not use this file except in compliance with the License.
-    You may obtain a copy of the License at
+You will likely stumble upon the continuous delivery webhook. Please do not modify or delete it.
```

Comment:
nit: remove. They might not stumble upon it. And if they modify it, they're on their own!

Reply:
fixed